lte/gateway/c/oai/tasks/amf/amf_sap.h

```diff
@@ -0,0 +1,214 @@
+  AMFREG_REGISTRATION_CNF   = _AMFREG_REGISTRATION_CNF,
+  AMFREG_DEREGISTRATION_REQ = _AMFREG_DEREGISTRATION_REQ,
+  AMFAS_DATA_REQ            = _AMFAS_DATA_REQ,
+  AMFCN_CS_RESPONSE         = AMFCN_SMC_PARAM_RES,
+};
+
+/*
+ * Minimal identifier for AMF-SAP primitives
+ */
+#define AMFREG_PRIMITIVE_MIN _AMFREG_START
+#define AMFAS_PRIMITIVE_MIN _AMFAS_START
+#define AMFCN_PRIMITIVE_MIN _AMFCN_START
+
+/*
+ * Maximal identifier for AMF-SAP primitives
+ */
+#define AMFREG_PRIMITIVE_MAX _AMFREG_END
+#define AMFAS_PRIMITIVE_MAX _AMFAS_END
+#define AMFCN_PRIMITIVE_MAX _AMFCN_END
+
+/****************************************************************************/
+/************************  G L O B A L    T Y P E S  ************************/
+/****************************************************************************/
+/*
+ * EMMREG primitive for registration  procedure
+ * -------------------------------------
+ */
+typedef struct amf_reg_register_s {
+  bool is_emergency; /* true if the UE was attempting to register to
+                      * the network for emergency services only  */
+  nas_amf_registration_proc_t* proc;
+} amf_reg_register_t;
+
+/*
+ * AMFREG primitive for de-registration procedure
+ * -------------------------------------
+ */
+typedef struct amf_reg_deregister_s {
+  bool switch_off; /* true if the UE is switched off       */
+  int type;        /* Network detach type              */
+} amf_reg_deregister_t;
+
+/*
+ * Structure of AMFREG-SAP primitive
+ */
+typedef struct sap_primitive_s {
+  amf_reg_register_t registered;
+  amf_reg_deregister_t deregister;
+  // amf_reg_sr_t sr;
+  amf_fsm_state_t previous_amf_fsm_state;
+  struct nas_amf_common_proc_s* common_proc;
+} sap_primitive_t;
+
+typedef struct amf_reg_s {
+  amf_primitive_t primitive;
+  amf_ue_ngap_id_t ue_id;
+  amf_context_t* ctx;
+  bool notify;  // notify through call-backs
+  bool free_proc;
+  sap_primitive_t u;
+} amf_reg_t;
+
+typedef struct amf_cn_auth_res_s {
+  /* UE identifier */
+  amf_ue_ngap_id_t ue_id;
+  /* For future use: nb of vectors provided */
+  // uint8_t nb_vectors;//TODO later
+
+} amf_cn_auth_res_t;
+
+typedef struct amf_cn_auth_fail_s {
+  /* UE identifier */
+  amf_ue_ngap_id_t ue_id;
+  /* NAS cause */
+  // nas_cause_t cause; //TODO
+} amf_cn_auth_fail_t;
+
+typedef struct amf_ul_s {
+  amf_cn_primitive_t primitive;
+  union {
+    amf_cn_auth_res_t* auth_res;
+    amf_cn_auth_fail_t* auth_fail;
+    // TODO many more structures to be defined based on Primitives
+  } u;
+} amf_cn_t;
+
+typedef struct primitive_s {
+  primitive_s() {}
+  ~primitive_s() {}
+  amf_reg_t amf_reg; /* AMFREG-SAP primitives    */
+  amf_as_t amf_as;   /* AMFAS-SAP primitives     */
+  amf_cn_t amf_cn;   /* AMFCN-SAP primitives     */
+} primitive_t;
+
+/*
+ * Structure of 5GMM Mobility Management primitive
+ */
+class amf_sap_t {
+  uint32_t count;
+
+ public:
+  amf_sap_t() {}
+  ~amf_sap_t() {}
+  amf_primitive_t primitive;
+  primitive_t u;
+};
```

Comment:
what is the purpose of each of these fields? I do not see any methods, looks like an overkill. Why not simply define this as a struct?

Reply:
changed to struct